src/Avalonia.Base/Input/GestureRecognizers/VelocityTracker.cs

```diff
@@ -0,0 +1,424 @@
+        ///
+        /// Larger values indicate greater quality.
+        public double Confidence { get; set; }
+    }
+
+    internal class LeastSquaresSolver
+    {
+        private const double PrecisionErrorTolerance = 1e-10;
+
+        /// <summary>
+        /// Fits a polynomial of the given degree to the data points.
+        /// When there is not enough data to fit a curve null is returned.
+        /// </summary>
+        public static PolynomialFit? Solve(int degree, ReadOnlySpan<double> x, ReadOnlySpan<double> y, ReadOnlySpan<double> w)
+        {
+            if (degree > x.Length)
+            {
+                // Not enough data to fit a curve.
+                return null;
+            }
+
+            PolynomialFit result = new PolynomialFit(degree);
+
+            // Shorthands for the purpose of notation equivalence to original C++ code.
+            int m = x.Length;
+            int n = degree + 1;
+
+            // Expand the X vector to a matrix A, pre-multiplied by the weights.
+            _Matrix a = new _Matrix(n, m);
+            for (int h = 0; h < m; h += 1)
+            {
+                a.Set(0, h, w[h]);
+                for (int i = 1; i < n; i += 1)
+                {
+                    a.Set(i, h, a.Get(i - 1, h) * x[h]);
+                }
+            }
+
+            // Apply the Gram-Schmidt process to A to obtain its QR decomposition.
+
+            // Orthonormal basis, column-major ordVectorer.
+            _Matrix q = new _Matrix(n, m);
+            // Upper triangular matrix, row-major order.
+            _Matrix r = new _Matrix(n, n);
+            for (int j = 0; j < n; j += 1)
+            {
+                for (int h = 0; h < m; h += 1)
+                {
+                    q.Set(j, h, a.Get(j, h));
+                }
+                for (int i = 0; i < j; i += 1)
+                {
+                    double dot = q.GetRow(j) * q.GetRow(i);
+                    for (int h = 0; h < m; h += 1)
+                    {
+                        q.Set(j, h, q.Get(j, h) - dot * q.Get(i, h));
+                    }
+                }
+
+                double norm = q.GetRow(j).Norm();
+                if (norm < PrecisionErrorTolerance)
+                {
+                    // Vectors are linearly dependent or zero so no solution.
+                    return null;
+                }
+
+                double inverseNorm = 1.0 / norm;
+                for (int h = 0; h < m; h += 1)
+                {
+                    q.Set(j, h, q.Get(j, h) * inverseNorm);
+                }
+                for (int i = 0; i < n; i += 1)
+                {
+                    r.Set(j, i, i < j ? 0.0 : q.GetRow(j) * a.GetRow(i));
+                }
+            }
+
+            // Solve R B = Qt W Y to find B. This is easy because R is upper triangular.
+            // We just work from bottom-right to top-left calculating B's coefficients.
+            _Vector wy = new _Vector(m);
```

Comment:
Isn't it possible to use [Vector](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.vector-1?view=net-7.0) type?
For netstandard2.0 this package can be referenced - [System.Numerics.Vectors](https://www.nuget.org/packages/System.Numerics.Vectors/) (if not already)

Reply:
@msneijders upd: I checked what can be done here. While it's [definitely possible](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvSasAShwB2GXtxYBhCPgAOvADY8AyjwBu/GEJGNmLSTLkwWASRk8IKnVH1hDw2iYnTZ82xj5SuPxGHmJm3pY2MgFBAiwAglD47qKm5j5WfjHBLAAaABxI7iIAzAzSuNgAZjBMpAwKNADeNAxt7W0qfLrYGLXMSAwAJhAcwJoMALIc6rIq6gCe5AAUWirYUgA8I2OaAHwMuuRoDGsb26PjMAe6pACUrR3tLdRPb8OXE7C4MxgMALwMOgsOjud5PKrQBjLXgyBi8AFAwTwhibQ6sAAyMCkAHMMAALZEIgDUgPID1e4OejypT2+vwYpPRAG1eABdBgAKkOpFZbLBtIYAF8aVTiAB2Bj02YCt4i6iijpdXg9PpiQY7K5TX68eYLUirdZbTX7dEnM7Gz7XHkU8EvQWHbBQQ4wMAYaBaXgAL1qgIAaq73VALrtropRjJZVSes7sGAwAG3VD/YHoCGrnsWAAtexR8ExhhVXhQXAYRNB3CIyYwfDQBaTJ24fHYdSKPAYdOaE7ltMm657ZZHO5594F3CuiBSIY9ktVmt1hsl5uthTtzswbup4N9vYD27DxXvSHOwdOlGA0EotFFktlre4FhY3EEonE4m2wX2h3tOMJreMwEb1LGdcD5LkGHHSApxAvkRzlQ83j7KVDAZQFTE4bg+DiGcWC0Lhll/GcD0pWlj2hAsEUA4tgPvR9sTxfFwN0LdPR9IlUXROjn0JeE3w/Wkv2/ZCflmACWXZJjeXZOCnnlBCnglYTfhk+UhSAA===) I have doubts if it actually makes much sense.